<h1 align="center">Python performance exercises</h1>

## Python best practices exercises

### Exercise 1

considering the following function for concatenating list strings with delimiter.

In [1]:
def ft_concatenate(l_strings, d):

    """concatenate list of strings into one string separated by delimiter"""

    res = l_strings[0]
    for e in l_strings[1:]:
        res = res + d + e
    return res

- profile the function and identify the bottlenecks.
- improve speed up of the function
*Hint: you may need to look to the string functions in python documentation*

In [114]:
# function profiling
d = " "
%prun ft_concatenate(["a","b","c"], d)



```
         4 function calls in 0.000 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 <ipython-input-1-1b2d70a09227>:1(ft_concatenate)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```



   

### Exercise 2

In this exercise you will solve the following problem using two methods bruteforce method, and fast method.

**Problem:** You are given a list of n integers, and your task is to calculate the number of distinct values in the list.

**Example**
- Input:
5
2 3 2 2 3

- Output:
2

**Implement the following methods:**

1. **bruteforce method:** create an empty list and start adding items for the given list without adding the previous item add, at the end the result list will contain unique values, print lenght of the list and you are done.
2. **fast method** think of using Set data structure.

- time the two methods, what do you think?

In [3]:
# bruteforce method
def bruteforce_method(l):
    result = []
    for num in l:
        if num not in result:
            result.append(num)
    return len(result)

In [4]:
# fast method
def fast_method(l):
    return len(set(l))

In [5]:
import random
# Create a random list of numbers for testing
test_list = [random.randint(0, 10) for _ in range(10)]
# Brute Force Method
%timeit bruteforce_method(test_list)
# Fast Method using Set
%timeit fast_method(test_list)

1.13 µs ± 332 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
360 ns ± 8.08 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## Cython exercises

### Exercise 1

1. load the cython extension.

In [6]:
%load_ext cython

2. Considering the following polynomial function:

In [7]:
def poly(a,b):
    return 10.5 * a + 3 * (b**2)

- Create an equivalent Cython function of `poly` with name `poly_cy`.

In [8]:
%%cython -a
cpdef float poly_cy(float a, float b):
    return 10.5 * a + 3 * (b*b)

3. time the performance of Python and Cython version of the function, what is the factor of speed up between the two verions.

In [13]:
%timeit poly(2,2)

494 ns ± 132 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [14]:
%timeit poly_cy(2,2)

100 ns ± 1 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


4. Now let's work on another example using loop.
    - rewrite the same function below fib that calculates the fibonacci sequence using cython, but now try to add type for the variables used inside it, add a prefix `_cy` to your new cython function.

In [15]:
def fib(n):
    a, b = 1, 1
    for i in range(n):
        a, b = a + b, a
    return a

In [16]:
%%cython -a
cpdef int fib_cy(int n):
    cdef int a=1, b=1,i
    for i in range(n):
        a, b = a + b, a
    return a

- time the two function for fibonacci series, with n = 20, what is the factor of speed now, What do you think?

In [17]:
%timeit fib(20)
%timeit fib_cy(20)

1.56 µs ± 343 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
102 ns ± 30.9 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


5. Recursive functions are functions that call themselves during their execution. Another interesting property of the Fibonacci sequence is that it can be written as a recursive function. That’s because each item depends on the values of other items (namely item n-1 and item n-2)

- Rewrite the fib function using recursion. Is it faster than the non-recursive version? Does Cythonizing it give even more of an advantage?

In [20]:
# Recursive Fibonacci
def fib_recursive(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib_recursive(n - 1) + fib_recursive(n - 2)

In [21]:
%timeit fib_recursive(20)

4.01 ms ± 1.27 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


The recursive version of fibonacci is clearly faster than the non recursive one

In [24]:
# Cython Recursive Fibonacci
%%cython -a
cpdef int fib_recursive_cy(int n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib_recursive_cy(n - 1) + fib_recursive_cy(n - 2)

In [25]:
%timeit fib_recursive_cy(20)

48.9 µs ± 9.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


### Exercise 2

- Monte Carlo methods are a broad class of computational algorithms that rely on repeated random sampling to obtain numerical results.
- One of the basic examples of getting started with the Monte Carlo algorithm is the estimation of Pi.

**Estimation of Pi**

- The idea is to simulate random (x, y) points in a 2-D plane with domain as a square of side 1 unit.
- Imagine a circle inside the same domain with same diameter and inscribed into the square.
- We then calculate the ratio of number points that lied inside the circle and total number of generated points.
- Refer to the image below:

![demo](../data/MonteCarloPlot.png)

We know that area of the square is 1 unit sq while that of circle is $\pi \ast  (\frac{1}{2})^{2} = \frac{\pi}{4}$. Now for a very large number of generated points,

![demo](../data/MonteCarloCalc.png)


## The Algorithm

1. Initialize cile_points, square_points and interval to 0.
2. Generate random point x.
3. Generate random point y.
4. Calculate d = x*x + y*y.
5. If d <= 1, increment circle_points.
6. Increment square_points.
7. Increment interval.
8. If increment < NO_OF_ITERATIONS, repeat from 2.
9. Calculate pi = 4*(circle_points/square_points).
10. Terminate.

**Your mission:** time the function `monte_carlo_pi`, identify the bottlenecks and create a new version using cython functionality to speed up monte carlo simulation for PI, use 100,000 points and compare the speed up factor between python and cython, considering the following optimizations:
- add type for variables used.
- add type for the function
- use c rand function instead of python rand function.

*Hint: you can import function from C libraries using the following approach `from libc.<name of c library> cimport <library function name>`, replace the holders `<>` with the right identities for the current problem*

In [109]:
import random
def monte_carlo_pi(nsamples):
    pi = 0
    circle_points = 0
    square_points = 0

    for i in range(nsamples):
        x = random.random()
        y = random.random()
        d = x**2 + y**2

        if d <= 1:
            circle_points += 1

        square_points += 1

        pi = 4 * (circle_points / square_points)
    return pi

In [111]:
%%cython -a
from libc.stdlib cimport rand, srand, RAND_MAX
def monte_carlo_pi_cy(int nsamples):

    srand(1)

    cdef int circle_points = 0
    cdef int square_points = 0
    cdef int interval = 0
    cdef int i
    cdef double x, y, d

    for i in range(nsamples):

        x = rand() / RAND_MAX
        y = rand() / RAND_MAX


        d = x * x + y * y

        if d <= 1:
            circle_points += 1
        square_points += 1
        interval += 1

    cdef double pi
    pi = 4.0 * (circle_points / square_points)

    return pi

In [115]:
%prun monte_carlo_pi(100000)



```
         200004 function calls in 0.105 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.088    0.088    0.105    0.105 <ipython-input-109-c20272d12438>:2(monte_carlo_pi)
   200000    0.017    0.000    0.017    0.000 {method 'random' of '_random.Random' objects}
        1    0.000    0.000    0.105    0.105 {built-in method builtins.exec}
        1    0.000    0.000    0.105    0.105 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```



In [116]:
%prun monte_carlo_pi_cy(100000)



```
         3 function calls in 0.006 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.006    0.006    0.006    0.006 <string>:1(<module>)
        1    0.000    0.000    0.006    0.006 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```



In [119]:
%timeit monte_carlo_pi(100000)
%timeit monte_carlo_pi_cy(100000)

79.7 ms ± 25.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
4.1 ms ± 51.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## Numba exercises

### Exercise 1

Previously we considered how to approximateby Monte Carlo.

- Use the same idea here, but make the code efficient using Numba.
- Compare speed with and without Numba when the sample size is large.

In [120]:
import numba as nb

@nb.jit
def monte_carlo_pi_nb(nsamples):
    pi = 0
    circle_points = 0
    square_points = 0

    for i in range(nsamples):
        x = random.random()
        y = random.random()
        d = x**2 + y**2

        if d <= 1:
            circle_points += 1

        square_points += 1

        pi = 4 * (circle_points / square_points)
    return pi

In [121]:
%timeit monte_carlo_pi_nb(100000)

1.38 ms ± 278 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


Obviously, using nb is very beneficial, we went from execution time of 79.7 ms to 1.38 ms using Numba

### Exercise 2

In the [Introduction to Quantitative Economics](https://python.quantecon.org/intro.html) with Python lecture series you can learn all about finite-state Markov chains.

For now, let's just concentrate on simulating a very simple example of such a chain.

Suppose that the volatility of returns on an asset can be in one of two regimes — high or low.

The transition probabilities across states are as follows ![markov](../data/markov.png)

For example, let the period length be one day, and suppose the current state is high.

We see from the graph that the state tomorrow will be

- high with probability 0.8

- low with probability 0.2

Your task is to simulate a sequence of daily volatility states according to this rule.

Set the length of the sequence to `n = 1_000_000` and start in the high state.

Implement a pure Python version and a Numba version, and compare speeds.

To test your code, evaluate the fraction of time that the chain spends in the low state.

If your code is correct, it should be about 2/3.

Hints:

- Represent the low state as 0 and the high state as 1.

- If you want to store integers in a NumPy array and then apply JIT compilation, use `x = np.empty(n, dtype=np.int_)`.


In [124]:
import numpy as np

def simulate_volatility_states(n):
    # Define the transition probabilities
    high_to_high_prob = 0.8
    high_to_low_prob = 0.2

    # Initialize the state to high (1)
    current_state = 1

    # Create an empty NumPy array to store the sequence of states as integers
    states = np.empty(n, dtype=np.int_)

    # Simulate the sequence
    for i in range(n):
        states[i] = current_state
        # Transition to the next state based on probabilities
        if current_state == 1:  # high state
            if random.random() < high_to_low_prob:
                current_state = 0  # low state
        else:  # low state
            if random.random() < high_to_high_prob:
                current_state = 1  # high state

    return states

In [126]:
# Calculate the fraction of time spent in the low state
@nb.jit
def simulate_volatility_states_nb(n):
    # Define the transition probabilities
    high_to_high_prob = 0.8
    high_to_low_prob = 0.2

    # Initialize the state to high (1)
    current_state = 1

    # Create an empty NumPy array to store the sequence of states as integers
    states = np.empty(n, dtype=np.int_)

    # Simulate the sequence
    for i in range(n):
        states[i] = current_state
        # Transition to the next state based on probabilities
        if current_state == 1:  # high state
            if random.random() < high_to_low_prob:
                current_state = 0  # low state
        else:  # low state
            if random.random() < high_to_high_prob:
                current_state = 1  # high state

    return states

In [129]:
states = simulate_volatility_states_nb(1000000)
np.mean(states == 0)

0.200299

In [127]:
%timeit simulate_volatility_states(1000000)
%timeit simulate_volatility_states_nb(1000000)

270 ms ± 11.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
11.3 ms ± 95.7 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [128]:
11.3/270

0.041851851851851855